nixos/modules/services/backup/sanoid.nix

```diff
@@ -0,0 +1,163 @@
+      type = types.lines;
+      default = "";
+    };
+  };
+
+  datasetOptions = commonOptions // {
+    useTemplate = mkOption {
+      description = "Names of the templates to use for this dataset";
+      type = types.nullOr (types.listOf (types.enum (attrNames cfg.templates)));
+      default = null;
+    };
+
+    recursive = mkOption {
+      description = "Whether to recursively snapshot dataset children";
+      type = types.nullOr (types.enum [ true false "zfs" ]);
+      default = null;
+    };
+
+    processChildrenOnly = mkOption {
+      description = "Whether to only snapshot child datasets if recursing";
+      type = types.nullOr types.bool;
+      default = null;
+    };
+  };
+
+  configBoolPrint = value: name: optionalString (value != null) "${name} = ${if value then "yes" else "no"}";
+  configPrint = value: name: optionalString (value != null) "${name} = ${toString value}";
+
+  configBlocks = blocks: prefix:
+    concatStringsSep "\n" (mapAttrsToList (d: v: ''
+      [${prefix}${d}]
+      ${configPrint v.hourly "hourly"}
+      ${configPrint v.daily "daily"}
+      ${configPrint v.monthly "monthly"}
+      ${configPrint v.yearly "yearly"}
+
+      ${configBoolPrint v.autoprune "autoprune"}
+      ${configBoolPrint v.autosnap "autosnap"}
+
+      ${optionalString ((v.useTemplate or null) != null) ''
+        use_template = ${concatStringsSep "," v.useTemplate}
+      ''}
+
+      ${configBoolPrint (v.recursive or null) "recursive"}
+      ${configBoolPrint (v.processChildrenOnly or null) "process_children_only"}
+      ${v.extraConfig}
+    '') blocks);
+
+  configFile = ''
+    ${configBlocks cfg.datasets ""}
+
+    ${configBlocks cfg.templates "template_"}
+  '';
+
+  configDir = pkgs.runCommand "sanoid-config" {} ''
+    mkdir -p "$out"
+    ln -s "${pkgs.writeText "sanoid.conf" configFile}" "$out/sanoid.conf"
+  '';
+
+in {
+
+    # Interface
+
+    options.services.sanoid = {
+      enable = mkEnableOption "Sanoid ZFS snapshotting service";
+
+      interval = mkOption {
+        type = types.str;
+        default = "hourly";
+        example = "daily";
+        description = ''
+          Run sanoid at this interval. The default is to run hourly.
+
+          The format is described in
+          <citerefentry><refentrytitle>systemd.time</refentrytitle>
+          <manvolnum>7</manvolnum></citerefentry>.
+        '';
+      };
+
+      datasets = mkOption {
+        type = types.attrsOf (types.submodule {
+          options = datasetOptions;
+        });
+        default = {};
+        description = "Datasets to snapshot";
+      };
+
+      templates = mkOption {
+        type = types.attrsOf (types.submodule {
+          options = commonOptions;
+        });
+        default = {};
+        description = "Templates for datasets";
+      };
+
+      extraArgs = mkOption {
+        description = "Extra arguments to pass to sanoid";
+        type = types.separatedString " ";
```

Comment:
`types.listOf types.str` is preferred, as it allows proper escaping of arguments, which you can then do with `lib.escapeShellArgs cfg.extraArgs` in `ExecStart`.

Same for the args in syncoid